pkg/controllers/node/node_syncer.go

```diff
@@ -39,12 +42,80 @@ func (c *NodeController) OnStatusUpdated(status bapi.SyncStatus) {
 	logrus.Infof("Node controller syncer status updated: %s", status)
 }
 
+func (c *NodeController) handleNewNode(u bapi.Update) {
+	if c.config.AutoHostEndpoints != "enabled" {
+		return
+	}
+
+	n := u.KVPair.Value.(*apiv3.Node)
+	// Check if we have a HEP created already for this node.
+	// If not, create one and log
+	// If we already have one, log anyways
+	_, err := c.calicoClient.HostEndpoints().Get(c.ctx, n.Name, options.GetOptions{})
+
+	if err != nil {
+		switch err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			logrus.Infof("host endpoint for node %q does not exist", n.Name)
+
+			// Copy over node labels to hep
+			hepLabels := make(map[string]string, len(n.Labels))
+			for k, v := range n.Labels {
+				hepLabels[k] = v
+			}
+
+			hep := &apiv3.HostEndpoint{
+				TypeMeta: metav1.TypeMeta{Kind: "HostEndpoint", APIVersion: "v3"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   n.Name,
+					Labels: hepLabels,
+				},
+				Spec: apiv3.HostEndpointSpec{
+					Node:          n.Name,
+					InterfaceName: "*",
+				},
+			}
+			h, err := c.calicoClient.HostEndpoints().Create(c.ctx, hep, options.SetOptions{})
+			// TODO: we should retry here a few times.
+			if err != nil {
+				logrus.Warnf("error creating host endpoint for new node %q: %v", n.Name, err.Error())
+				return
+			}
+			logrus.Infof("created hep for new node %q", n.Name)
+			return
+		default:
+			logrus.Infof("error getting hep: %v", err.Error())
+			return
+		}
+	}
+	logrus.Infof("host endpoint for node %q already exists", n.Name)
+}
+
+func (c *NodeController) deleteHEP(nodeName string) {
+	if c.config.AutoHostEndpoints != "enabled" {
+		return
+	}
+
+	h, err := c.calicoClient.HostEndpoints().Delete(c.ctx, n, options.DeleteOptions{})
+	// TODO: retry
+	if err != nil {
+		switch err.(type) {
+		case errors.ErrorResourceDoesNotExist:
+			logrus.Warnf("could not delete host endpoint for node %q because it does not exist", n.Name)
+			return
+		default:
+			// TODO: retry for other errors
+		}
+	}
+}
+
 func (c *NodeController) OnUpdates(updates []bapi.Update) {
 	logrus.Debugf("Node controller syncer received updates: %#v", updates)
 	for _, upd := range updates {
 		switch upd.UpdateType {
 		case bapi.UpdateTypeKVNew:
-			fallthrough
```

Comment:
I think we want to treat KVNew and KVUpdated more or less the same - for example, we want to update the labels on the HEP if the labels on the node change. The simplest way to do that is just to have common logic for both cases.